Sources/NIOHTTP1/HTTPDecoder.swift

```diff
@@ -357,92 +377,174 @@ public class HTTPDecoder<HTTPMessageT>: ByteToMessageDecoder, AnyHTTPDecoder {
+    // Decode HTTP until there is nothing more to decode.
+    private func decodeHTTP(ctx: ChannelHandlerContext) throws {
+        // We need to refetch the cumulationBuffer on each loop as it may has changed due re-entrance calls of channelRead(...)
+        while let bufferSlice = self.cumulationBuffer, bufferSlice.readableBytes > 0 {
+            // Using withVeryUnsafeBytes here as this simplifies the calculation of the readerIndex which is relative to the baseAddress.
+            let result = bufferSlice.withVeryUnsafeBytes { (pointer) -> size_t in
+                self.state.baseAddress = pointer.baseAddress!.assumingMemoryBound(to: UInt8.self)
+                defer {
+                    self.state.baseAddress = nil
+                }
+                return self.state.baseAddress!.withMemoryRebound(to: Int8.self, capacity: pointer.count) { p in
+                    c_nio_http_parser_execute(&self.parser, &self.settings, p.advanced(by: bufferSlice.readerIndex), bufferSlice.readableBytes)
+                }
+            }
+            try self.rethrowParserError()
+
+            /// http_parser_execute(...) should always consume the whole buffer.
+            assert(result == bufferSlice.readableBytes)
+
+            // Update readerIndex of the cumulationBuffer itself as we will refetch it in the next loop run if needed.
+            self.cumulationBuffer?.moveReaderIndex(forwardBy: result)
+        }
+
+        if self.state.seenEOF {
+            // We need to notify the parser about the EOF as we received it while in http_parser_execute.
+            self.notifyParserEOF(ctx: ctx)
+        }
+    }
+
+    private func discardDecodedBytes() {
+        guard self.cumulationBuffer != nil else {
+            // Guard against the case of closing the channel. In this case the cumulationBuffer will be nil.
+            return
         }
 
-        let result = try buffer.withVeryUnsafeBytes { (pointer) -> size_t in
-            state.baseAddress = pointer.baseAddress!.assumingMemoryBound(to: UInt8.self)
+        assert(self.cumulationBuffer!.readableBytes == 0)
+
+        switch self.state.dataAwaitingState {
+        case .body, .messageBegin:
+            assert(self.state.currentNameIndex == nil)
+            assert(self.state.currentHeaders.isEmpty)
+            assert(self.state.slice == nil)
 
-            let result = state.baseAddress!.withMemoryRebound(to: Int8.self, capacity: pointer.count, { p in
-                c_nio_http_parser_execute(&parser, &settings, p.advanced(by: buffer.readerIndex), buffer.readableBytes)
-            })
+            // Its safe to just drop the cumulationBuffer as we not have any extra views into it that are represented as readerIndex / length.
+            self.cumulationBuffer = nil
 
-            if let error = state.currentError {
-                throw error
+        case .headerField, .headerValue:
+            guard let headerStartIdx = self.state.headerStartIndex else {
```

Comment:
Is it valid to be in this state *without* a `headerStartIndex`?

Reply:
@Lukasa yes as we reset it to nil if we discarded the decoded bytes.

Reply:
Is there any reason not to set it to 0 instead and avoid this branch?

Reply:
Yes as otherwise we will try to increase the writerIndex etc later on which is even more expensive.